Java/core/src/main/java/com/amazon/randomcutforest/PredictiveRandomCutForest.java

```diff
@@ -0,0 +1,377 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.amazon.randomcutforest;
+
+import static com.amazon.randomcutforest.CommonUtils.checkArgument;
+import static com.amazon.randomcutforest.CommonUtils.toDoubleArray;
+import static com.amazon.randomcutforest.RandomCutForest.DEFAULT_BOUNDING_BOX_CACHE_FRACTION;
+import static com.amazon.randomcutforest.RandomCutForest.DEFAULT_CENTER_OF_MASS_ENABLED;
+import static com.amazon.randomcutforest.RandomCutForest.DEFAULT_INITIAL_ACCEPT_FRACTION;
+import static com.amazon.randomcutforest.RandomCutForest.DEFAULT_NUMBER_OF_TREES;
+import static com.amazon.randomcutforest.RandomCutForest.DEFAULT_OUTPUT_AFTER_FRACTION;
+import static com.amazon.randomcutforest.RandomCutForest.DEFAULT_PARALLEL_EXECUTION_ENABLED;
+import static com.amazon.randomcutforest.RandomCutForest.DEFAULT_SAMPLE_SIZE;
+import static com.amazon.randomcutforest.RandomCutForest.DEFAULT_SHINGLE_SIZE;
+import static com.amazon.randomcutforest.RandomCutForest.DEFAULT_STORE_SEQUENCE_INDEXES_ENABLED;
+import static com.amazon.randomcutforest.config.ImputationMethod.PREVIOUS;
+import static com.amazon.randomcutforest.preprocessor.Preprocessor.DEFAULT_START_NORMALIZATION;
+import static com.amazon.randomcutforest.preprocessor.Preprocessor.DEFAULT_STOP_NORMALIZATION;
+import static java.lang.Math.max;
+import static java.lang.Math.min;
+
+import java.util.Arrays;
+import java.util.Optional;
+import java.util.Random;
+
+import lombok.Getter;
+import lombok.Setter;
+
+import com.amazon.randomcutforest.config.ForestMode;
+import com.amazon.randomcutforest.config.ImputationMethod;
+import com.amazon.randomcutforest.config.TransformMethod;
+import com.amazon.randomcutforest.preprocessor.IPreprocessor;
+import com.amazon.randomcutforest.preprocessor.Preprocessor;
+import com.amazon.randomcutforest.returntypes.SampleSummary;
+
+/**
+ * This class provides a predictive imputation based on RCF (respecting the
+ * arrow of time) alongside streaming normalization
+ */
+@Getter
+@Setter
+public class PredictiveRandomCutForest {
```

Comment:
You might have missed my comment.

Reply:
No :) Stay tuned. The vision is PredictiveRCF to be a combination of Preprocessor + RandomCutForest. The current RandomCutForest class is backward compatible with RCF 1.0 (state only), 2.0, 3.0 ... changing that class to accommodate preprocessing seems daunting at the moment (may be feasible down the road), and we do not want to break the backward compatibility. PredictiveRCF would focus on the higher level and the existing RandomcutForest will focus on the more detail focused level. In fullness of time there can be convergence -- note that this is how we went from RCF1.0 (pointer trees) to RCF2.0 (compact RCF) by slowly changing in place. This would hopefully avoid any regression/unwanted/unexplained changes.

Reply:
Take a look at ErrorHandler -- it can now use simple statistics or a more involved predictive model (PredictiveRCF) where we predict the last two columns of (current value, lookahead, _positive_error, _negative_error). Overall the estimation is better for shorter lookaheads -- but does take more space and time. It is an option and can be built upon. Also in the fullness of time ThresholdedRCF can use PredicitiveRCF directly and avoid messing with transformations. Right now ParkServices correspond to time series analysis (thresholding corresponds to a sequence of values, same for forecasts) and do not have flexibility of updates (because it is difficult to turn back on) whereas the core packages the flexibility of updates.